designs/odcr.md

```diff
@@ -0,0 +1,204 @@
+- Supporting Capacity Reservation Groups. _Adding this abstraction for now adds additional complexity_
+- Supporting updating Capacity Reservations availabile instance count after success of CreateFleet API. _We will currenlty rely on reconsolidation of capacity reservations itself, and rely on CreateFleet API throwing a `ReservationCapacityExceeded`_
+- Supporting conditions for capacity reservation instance creations. _By showing the available instance count within each capacity reservation for an EC2NodeClass we have similar information available_
+
+## Proposed Solution
+
+### Supporting associating Capacity Reservations to EC2NodeClass
+
+- Add a new field under `spec` for `capacityReservationSelectorTerms` to `EC2NodeClass` for defining which Capacity Reservation to be used for a specific `EC2NodeClass`
+- Add a new field under `status` for the found Capacity Reservations by the `spec.capacityReservationSelectorTerms` for the `EC2NodeClass`
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSelectorTerms:
+    - # The Availability Zone of the Capacity Reservation
+      availabilityZone: String | None
+      # The platform of operating system for which the Capacity Reservation reserves capacity
+      id: String | None
+      # The type of operating system for which the Capacity Reservation reserves capacity
+      instancePlatform: String | None
+      # The type of operating system for which the Capacity Reservation reserves capacity
+      instanceType: String | None
+      # The ID of the Amazon Web Services account that owns the Capacity Reservation
+      ownerId: String | None
+      # Tags is a map of key/value tags used to select subnets
+      # Specifying '*' for a value selects all values for a given tag key.
+      tags: Map | None
+      # Indicates the tenancy of the Capacity Reservation.
+      # A Capacity Reservation can have one of the following tenancy 'default' or 'dedicated':
+      #   default - The Capacity Reservation is created on hardware that is shared with other Amazon Web Services accounts.
+      #   dedicated - The Capacity Reservation is created on single-tenant hardware that is dedicated to a single Amazon Web Services account.
+      tenancy: String | None
+status:
+  capacityReservations:
+    - # AvailabilityZone of the Capacity Reservation
+      availabilityZone: String
+      # Available Instance Count of the Capacity Reservation
+      availableInstanceCount: Integer
+      # ID of the Capacity Reservation
+      id: String
+      # InstanceType of the Capacity Reservation
+      instanceType: String
+      # Total Instance Count of the Capacity Reservation
+      totalInstanceCount: Integer
+```
+
+This follows closely (does not implement all fields) how EC2 [DescribeCapacityReservations](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html) can filter.
```

Comment:
I assume this is in reference to the selector terms? So I assume you're purposely excluding instance-match-criteria,  placement-group-arn, and start/end dates?

Reply:
instanceMatchCriteria is only needed if we support something outside of target for this first implementation

Reply:
I dont see a use case where you filter based on start/end date for the selector terms, as we by default only filter for `active` ones